File: kakao.app/js/ked/util/syntax.js
var _k_ = {empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, isArr: function (o) {return Array.isArray(o)}, trim: function (s,c=' ') {return _k_.ltrim(_k_.rtrim(s,c),c)}, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, ltrim: function (s,c=' ') { while (_k_.in(s[0],c)) { s = s.slice(1) } return s}, rtrim: function (s,c=' ') {while (_k_.in(s.slice(-1)[0],c)) { s = s.slice(0, s.length - 1) } return s}}

var syntax

import matchr from "../../kxk/matchr.js"
import kstr from "../../kxk/kstr.js"
import kseg from "../../kxk/kseg.js"

import kulur from "../../kolor/kulur.js"

import theme from "../theme/theme.js"


syntax = (function ()
{
    function syntax (name)
    {
        this.name = name
    
        this["getChar"] = this["getChar"].bind(this)
        this["getColor"] = this["getColor"].bind(this)
        this["getClass"] = this["getClass"].bind(this)
        this["appendSegls"] = this["appendSegls"].bind(this)
        this["addSegl"] = this["addSegl"].bind(this)
        this["setSegls"] = this["setSegls"].bind(this)
        this["setLines"] = this["setLines"].bind(this)
        this["setRgxs"] = this["setRgxs"].bind(this)
        this["setExt"] = this["setExt"].bind(this)
        this["clear"] = this["clear"].bind(this)
        this.ext = 'txt'
        this.clear()
    }

    syntax.prototype["clear"] = function ()
    {
        this.diss = []
        this.hash = {}
        return this.liha = {}
    }

    syntax.prototype["setExt"] = function (ext)
    {
        this.ext = ext
    }

    syntax.prototype["setRgxs"] = function (rgxs)
    {
        return this.config = matchr.config(rgxs,'u')
    }

    syntax.prototype["setLines"] = function (lines)
    {
        return this.setSegls(kseg.segls(lines))
    }

    syntax.prototype["setSegls"] = function (segls)
    {
        var ds, dss, hsh, idx, segl, segs

        if (!_k_.empty(this.config))
        {
            this.diss = []
            var list = _k_.list(segls)
            for (var _a_ = 0; _a_ < list.length; _a_++)
            {
                segs = list[_a_]
                dss = matchr.ranges(this.config,kseg.str(segs),'u')
                var list1 = _k_.list(dss)
                for (var _b_ = 0; _b_ < list1.length; _b_++)
                {
                    ds = list1[_b_]
                    if (_k_.isArr(ds.clss))
                    {
                        ds.clss = _k_.trim(ds.clss.map(kstr.trim).join(' '))
                    }
                }
                this.diss.push(dss)
            }
        }
        else
        {
            if (this.partialUpdate(segls))
            {
                return
            }
            this.clear()
            this.diss = kulur.dissect(segls,this.ext)
            var list2 = _k_.list(segls)
            for (idx = 0; idx < list2.length; idx++)
            {
                segl = list2[idx]
                hsh = kseg.hash(segl)
                this.hash[hsh] = this.diss[idx]
                this.liha[idx] = hsh
            }
        }
    }

    syntax.prototype["partialUpdate"] = function (segls)
    {
        var hsh, idx, newHash, segl

        if (this.diss.length !== segls.length)
        {
            return
        }
        if (_k_.empty(this.hash))
        {
            return
        }
        newHash = {}
        var list = _k_.list(segls)
        for (idx = 0; idx < list.length; idx++)
        {
            segl = list[idx]
            hsh = kseg.hash(segl)
            if (this.hash[hsh])
            {
                newHash[hsh] = this.hash[hsh]
            }
            else
            {
                newHash[hsh] = kulur.dissect([segl],this.ext)[0]
            }
            if (this.liha[idx] !== hsh)
            {
                this.diss.splice(idx,1,newHash[hsh])
            }
        }
        return this.hash = newHash
    }

    syntax.prototype["addSegl"] = function (segl, ext)
    {
        return this.diss = this.diss.concat(kulur.dissect([segl],ext))
    }

    syntax.prototype["appendSegls"] = function (segls, ext)
    {
        var segl

        var list = _k_.list(segls)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            segl = list[_a_]
            this.addSegl(segl,ext)
        }
    }

    syntax.prototype["getClass"] = function (x, y)
    {
        var dss

        var list = _k_.list(this.diss[y])
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            dss = list[_a_]
            if ((dss.start <= x && x < dss.start + dss.length))
            {
                return dss.clss
            }
        }
        return 'text'
    }

    syntax.prototype["getColor"] = function (x, y)
    {
        var clss, _96_27_

        if (_k_.isNum(x))
        {
            clss = this.getClass(x,y)
        }
        else
        {
            clss = x
        }
        return ((_96_27_=theme.syntax[clss]) != null ? _96_27_ : '#ff0000')
    }

    syntax.prototype["getChar"] = function (x, y, char)
    {
        var clss

        clss = this.getClass(x,y)
        if (_k_.in('header',clss))
        {
            return '█'
        }
        return char
    }

    return syntax
})()

export default syntax;

File: kakao.app/love/lua/util/syntax.lua
--[[
 0000000  000   000  000   000  000000000   0000000   000   000
000        000 000   0000  000     000     000   000   000 000 
0000000     00000    000 0 000     000     000000000    00000  
     000     000     000  0000     000     000   000   000 000 
0000000      000     000   000     000     000   000  000   000
--]]


local syntax = class("syntax")
    


function syntax:init(name) 
        self.name = name
        self.ext = 'txt'
        self:clear()
        return self
    end


function syntax:clear() 
        self.diss = array()
        self.hash = {}
        self.liha = {}
        return self.liha
    end


function syntax:setExt(ext) 
            self.ext = ext ; self.ansi = false ; return self.ansi
    end


function syntax:setRgxs(rgxs) 
        self.config = matchr.config(rgxs)
        return -- log "setRgxs" @config
    end


function syntax:setLines(lines) 
    return self:setSegls(kseg.segls(lines))
    end


function syntax:setColors(colors) 
               self.colors = colors ; self.ansi = true ; return self.ansi
    end

--  ███████  ████████   ███████   ███       ███████
-- ███       ███       ███        ███      ███     
-- ███████   ███████   ███  ████  ███      ███████ 
--      ███  ███       ███   ███  ███           ███
-- ███████   ████████   ███████   ███████  ███████ 


function syntax:setSegls(segls) 
        if valid(self.config) then 
            self.diss = array()
            for _, segs in ipairs(segls) do 
                local dss = matchr.ranges(self.config, kseg.str(segs))
                print("SEGS", kseg.str(segs))
                print("DSS", dss)
                for _, ds in ipairs(dss) do 
                    if is(ds.clss, array) then 
                        ds.clss = trim(ds.clss:map(kstr.trim).join(' '))
                    end
                end
                
                self.diss.push(dss)
            end
            
            return -- log "#{@name} diss" @diss if @name == "funcol_funtree.state.syntax"
        else 
            -- ⮐  if @partialUpdate segls
            
            self:clear()
            
            self.diss = kulur.dissect(segls, self.ext)
            -- for idx segl in segls
            --     hsh = kseg.hash(segl)
            --     @hash[hsh] = @diss[idx]
            return --     @liha[idx] = hsh
        end
    end

-- ███   ███  ████████   ███████     ███████   █████████  ████████
-- ███   ███  ███   ███  ███   ███  ███   ███     ███     ███     
-- ███   ███  ████████   ███   ███  █████████     ███     ███████ 
-- ███   ███  ███        ███   ███  ███   ███     ███     ███     
--  ███████   ███        ███████    ███   ███     ███     ████████


function syntax:partialUpdate(segls) 
        if (self.diss.length ~= segls.length) then return end
        if empty(self.hash) then return end
        
        local newHash = {}
        
        for segl, idx in ipairs(segls) do 
            local hsh = kseg.hash(segl)
            if self.hash[hsh] then 
                newHash[hsh] = self.hash[hsh]
            else 
                newHash[hsh] = kulur.dissect(array(segl), self.ext)[0]
            end
            
            if (self.liha[idx] ~= hsh) then 
                self.diss.splice(idx, 1, newHash[hsh])
                self.liha[idx] = hsh
            end
        end
        
        self.hash = newHash
        return self.hash
    end


function syntax:addSegl(segl, ext) 
        self.diss = self.diss.concat(kulur.dissect(array(segl), ext))
        return self.diss
    end


function syntax:appendSegls(segls, ext) 
        -- @diss = @diss.concat kulur.dissect(segls ext)
        for _, segl in ipairs(segls) do 
            self:addSegl(segl, ext)
        end
    end

--  ███████  ███       ███████    ███████   ███████
-- ███       ███      ███   ███  ███       ███     
-- ███       ███      █████████  ███████   ███████ 
-- ███       ███      ███   ███       ███       ███
--  ███████  ███████  ███   ███  ███████   ███████ 


function syntax:getClass(x, y) 
        if (self.diss and self.diss[y]) then 
            for dss in self.diss[y]:each() do 
                if ((dss.start <= x) and (x < (dss.start + dss.length))) then 
                    -- log "DISS " dss.start, x, dss.start+dss.length, dss.clss
                    return dss.clss
                end
            end
        end
        
        return 'text'
    end

--  ███████   ███████   ███       ███████   ████████ 
-- ███       ███   ███  ███      ███   ███  ███   ███
-- ███       ███   ███  ███      ███   ███  ███████  
-- ███       ███   ███  ███      ███   ███  ███   ███
--  ███████   ███████   ███████   ███████   ███   ███


function syntax:getColor(x, y) 
        if self.ansi then return self:getAnsiColor(x, y) end
        local clss = x
        if not is(clss, "string") then 
            clss = self:getClass(x, y)
        end
        
        -- if empty theme.syntax[clss] and not clss.endsWith 'italic'
        
            -- log "syntax.getColor - no syntax color for '#{clss}'" # @diss[y] 
        -- if clss is "string"
        --     log "COLOR #{clss}" theme.syntax[clss] 
        if theme.syntax[clss] then 
            -- log type(theme.syntax[clss])
            -- log theme.syntax[clss][1]
            -- log theme.syntax[clss][2]
            -- log theme.syntax[clss][3]
            return theme.syntax[clss]
        else 
            return array(155, 155, 155)
        end
    end

--  ███████   ███   ███   ███████  ███
-- ███   ███  ████  ███  ███       ███
-- █████████  ███ █ ███  ███████   ███
-- ███   ███  ███  ████       ███  ███
-- ███   ███  ███   ███  ███████   ███


function syntax:getAnsiColor(x, y) 
        for clr in self.colors[y] do 
            if (clr.x <= x) then 
                if clr.w then 
                    if (x < (clr.x + clr.w)) then 
                        return clr.fg
                    end
                else 
                    return clr.fg
                end
            end
        end
        
        return array(256, 0, 256)
    end

--  ███████  ███   ███   ███████   ████████ 
-- ███       ███   ███  ███   ███  ███   ███
-- ███       █████████  █████████  ███████  
-- ███       ███   ███  ███   ███  ███   ███
--  ███████  ███   ███  ███   ███  ███   ███


function syntax:getChar(x, y, char) 
        local clss = self:getClass(x, y)
        if (kstr.find(clss, 'header') >= 1) then return '█' end
        return char
    end

return syntax